incubator/hnc/pkg/controllers/object_controller.go

```diff
@@ -104,68 +103,93 @@ func (r *ObjectReconciler) SyncNamespace(ctx context.Context, log logr.Logger, n
+		err := r.Update(ctx, inst)
+		if err != nil {
+			log.Error(err, "Couldn't add modified annotation")
+		}
+		return nil, true
 	}
 
-	// If there's no source, *this* is the source
+	return r.Forest.Get(inst.GetNamespace()).ChildNames(), true
+}
+
+// hasCorrectAncestry checks to see if the given object has correct ancestry. It returns false
+// if the source namespace is no longer an ancestor and the object should be deleted.
+func (r *ObjectReconciler) hasCorrectAncestry(ctx context.Context, log logr.Logger, inst, src *unstructured.Unstructured) bool {
 	if src == nil {
-		return true, nil
+		log.Info("Will delete because the source instance no longer exists", "inheritedFrom", getSourceNS(inst))
+		return false
+	}
+
+	if inst == src {
+		return true
+	}
+
+	// The source exists. Is it still in an ancestor namespace?
+	curNS := r.Forest.Get(inst.GetNamespace())
+	srcNS := r.Forest.Get(src.GetNamespace())
+	if !curNS.IsAncestor(srcNS) {
+		log.Info("Will delete because the source namespace is no longer an ancestor", "inheritedFrom", srcNS)
+		return false
+	}
+
+	return true
+}
+
+// isUnmodified checks to see if this object is the same as the source. It returns true if it's
+// correct and ready to be propagated.
+func (r *ObjectReconciler) isUnmodified(ctx context.Context, log logr.Logger, inst, src *unstructured.Unstructured) bool {
+	if src == inst {
+		log.Info("This instance is the src itself")
+		return true
 	}
 
-	// We found the object in an ancestor namespace. Is this object the same as the source?
 	instCopied := copyObject(inst)
 	// The only difference could be namespace and labelInheritedFrom label.
 	// Make the namespaces and inheritedFrom labels the same before comparison.
 	instCopied.SetNamespace(src.GetNamespace())
 	setLabel(instCopied, labelInheritedFrom, "")
 	srcCopied := copyObject(src)
 	setLabel(srcCopied, labelInheritedFrom, "")
-	equal := reflect.DeepEqual(instCopied, srcCopied)
-	if !equal {
-		log.Info("Marking as modified from the source", "inheritedFrom", src.GetNamespace())
-		// Mark as modified if the copied fields don't match.
-		setAnnotation(inst, annotationModified, "true")
-		err := r.Update(ctx, inst)
-		if err != nil {
-			log.Error(err, "Couldn't add modified annotation")
-			return false, err
-		}
-	}
-	return equal, nil
+	return reflect.DeepEqual(instCopied, srcCopied)
 }
 
-// getSourceInst gets source instance of this one. It returns nil if this instance isn't a copy.
-// It returns true if the source instance is missing. It could be that the source instance no
-// longer exists or the source namespace is no longer an ancestor.
-func (r *ObjectReconciler) getSourceInst(ctx context.Context, log logr.Logger, inst *unstructured.Unstructured) (bool, *unstructured.Unstructured, error) {
+// getSourceInst gets source instance of this one. It returns itself if it's the source. It returns
+// nil if the source instance no longer exists in the source namespace or there's an error.
```

Comment:
I think this method needs to return an error - we want to return an error when we want the reconciler to try again, and "I temporarily couldn't access the apiserver" is probably a very good example of that.

If there's a temporary interruption accessing the apiserver and we delete this object, we'll never revisit it!

So this should return (*unstructured, error), if if the caller gets an error, it should pass that on to the top level.

Reply:
Done. Changed it back to return error when it's not `IsNotFound` error.